### DISCUSSION TITLE
Enforcing Uniqueness of a Collection

### DISCUSSION THREAD

**Discussion by jsormaz** [sqlalchemy/sqlalchemy#10860] [Usage Questions] [answered]
Let's say I have 2 classes with a many to many constraint between them.
Is there a clever way I can enforce uniqueness of the collection?

some pseudocode to give the idea:
```python
class AssemblyPartAssociation(Base):
    part_id = Column(Integer, ForeignKey("Part.id"))
    assembly_id = Column(Integer, ForeignKey("Assembly.id"))

class Assembly(Base):
    parts = relationship('Part', secondary="assembly_part_association")

class Part(Base):
    assemblies = relationship(Assembly, secondary="assembly_part_association")
```

Is there a clever way I can enforce uniqueness of the collection?

for example,
```python
Assembly(id=1 parts=[Part(id=1), Part(id=2)])
Assembly(id=2 parts=[Part(id=1), Part(id=2)])
``` 
should cause an integrity error,
but 
```python
Assembly(id=1 parts=[Part(id=1), Part(id=2)])
Assembly(id=2 parts=[Part(id=1), Part(id=2), Part(id=3)])
```
would be valid

There should not be another Assembly consisting of the same combination of parts, but there might be a different assembly using additional parts or a subset of the parts.

**Comment by zzzeek** [comment]
you would use `collection_class=set` most easily

https://docs.sqlalchemy.org/en/20/orm/collection_api.html#custom-collections

**Reply by jsormaz** [reply]
hmm, this is cool and something new for me but I'm not sure that is what I am after. That would still allow multiple parents to have the same set of child items.

I want to enforce that for a given set of child items, there is only 1 parent that contains that set.

**Reply by CaselIT** [reply, marked as answer]
no that's not something that's implemented.
I also don't think that this is something that can be enforced on the db side.

**Reply by jsormaz** [reply]
Thanks, I kind of figured it would be something I had to enforce in my business logic.

I appreciate the quick replies.
